pmd/dependencies.bzl

```diff
@@ -26,15 +40,19 @@ def rules_pmd_dependencies():
         sha256 = rules_java_sha,
     )
 
-    # JVM External
-
-    rules_jvm_external_version = "4.2"
-    rules_jvm_external_sha = "2cd77de091e5376afaf9cc391c15f093ebd0105192373b334f0a855d89092ad5"
```

Comment:
This rule is as simple as it gets if I remember right — there are no unit tests and no worker support. While this change might work out now, I imagine it will block both unit tests and worker — unless both are implemented without dependencies.

Do you have examples how other rules handle this? I imagine we can do pinned dependencies but it might be a PITA if consumers want to change the PMD version (the pin file should be re-generated and somehow passed to the rule).

Reply:
> While this change might work out now, I imagine it will block both unit tests and worker — unless both are implemented without dependencies.

This shouldn't change any of that, as we are just pulling the PMD jars from a zip file instead of rules_jvm_external. When you include and invoke these rules Bazel will automatically compile these rules against what ever version of PMD you give it.

The thing that can break, which is also the case using the implementation in master, is that PMD could chance it's implementation in a way that breaks how `Main.java` is referencing it. If when this happens these rules just wouldn't work.

> Do you have examples how other rules handle this?

`rules_kotlin` is one example. It uses a custom repository to dynamically pass in what ever version of rules_kotlin we support, which is what this PR is doing. The main difference is that rules_kotlin ships all of it's wrapper code as a pre-compiled binary that side loads + uses reflection to invoke the Kotlin compiler.

> I imagine we can do pinned dependencies but it might be a PITA if consumers want to change the PMD version (the pin file should be re-generated and somehow passed to the rule).

Yeah, pinning these dependencies wouldn't be easy to do. `rules_jvm_external` generates a custom repository for you underneath the hood that you have to load a bzl file from when pinning, and we have no way to re-pin dynamically if our users decide to change PMD releases.

Reply:
> This shouldn't change any of that, as we are just pulling the PMD jars from a zip file instead of rules_jvm_external. When you include and invoke these rules Bazel will automatically compile these rules against what ever version of PMD you give it.

True but I was talking about a bit different thing. For example, Detekt rules pull RxJava for the worker support and JUnit for unit tests. I imagine that avoiding JVM External leads to pulling all of that and transitive dependencies _or_ avoiding all dependencies altogether.

Reply:
Ah yes I see what you mean. For this use case we have a few options:
1. Go down a more complex path and switch to pre-compiled releases where we side load pmd/detekt jars into the classpath
2. Be extra strict about our policy on pulling in external dependencies and use `http_jar` to fetch just what we need instead of relying on `rules_jvm_external` to fetch the world.
3. We can take a more hybrid approach and compile the generic worker implementation into something that can be imported as a standalone jar or even checked into the repo directly. This approach would be a hybrid of option 1 and 2.